Implement a Collins Propagator

em-archer opened the pull request:
There is now a basic implementation of the Collins propagator for propagating beams to the focus. In the MWE below, a 150mm focal length lens is used.

Current limitations: 
- Due to an issue where I couldn't reallocate different length axes to ``laser.grid``, the automatic scaling of the grids based on a desired magnification isn't currently implemented. This means that the user needs to tune the number of gridpoints and the width of the grid at the input plane for the desired resolution of the focus. Will fix using padding in future.
- Chromatic optics not yet supported, although this only requires modifying ``add_lens`` to tile a spectral dependence across all spatial grid points. 
- Haven't checked whether it works for lenses with negative focal lengths. At the moment, the user MUST specify a lens.
- Need to check energy conservation for correct electric field strength.

```py
from lasy.laser import Laser
from lasy.profiles.gaussian_profile import GaussianProfile
from lasy.propagators import CollinsSFFTPropagator
import numpy as np
c = 2.998e8

wavelength = 800e-9  # Laser wavelength in meters
polarization = (1, 0)  # Linearly polarized in the x direction
energy = 1.5  # Energy of the laser pulse in joules
spot_size = 6.0e-3  # Waist of the laser pulse in meters
pulse_duration = 30e-15  # Pulse duration of the laser in seconds
t_peak = 0.0  # Location of the peak of the laser pulse in time

laser_profile = GaussianProfile(
    wavelength, polarization, energy, spot_size, pulse_duration, t_peak
)

dimensions = "xyt"  # Use cylindrical geometry
lo = (-5. * spot_size, -5. * spot_size, -2.5 * pulse_duration)  # Lower bounds of the simulation box
hi = ( 5. * spot_size,  5. * spot_size,  2.5 * pulse_duration)  # Upper bounds of the simulation box
num_points = (100, 100, 500)  # Number of points in each dimension

laser = Laser(dimensions, lo, hi, num_points, laser_profile)
laser.show()
```
<img width="628" alt="image" src="https://github.com/user-attachments/assets/71892518-2202-472c-8e23-29613c782f8c" />


```
z_R = np.pi * spot_size**2 / wavelength  # The Rayleigh length
f_0 = 150e-3
omega0 = 2 * np.pi * c / 0.8e-6
dim = 'xyt'

propagator = CollinsSFFTPropagator(omega0, dim)
propagator.add_lens(f_0) # Add a lens
propagator.add_vacuum(f_0) # Go to focus
propagator.add_vacuum(-3.*150e-6) # A few Rayleigh ranges away
propagator.propagate(laser.grid) # Propagate the pulse upstream of the focal plane
laser.show()

```
<img width="652" alt="image" src="https://github.com/user-attachments/assets/21e2186b-d96a-4f07-bd4e-934310c3ffbe" />

MaxThevenet committed:
skeleton code for propagators

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

MaxThevenet committed:
add fft wrapper function

Commit:
Modifying Collins propagator

Commit:
.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

em-archer committed:
Merge branch 'LASY-org:development' into collins

MaxThevenet committed:
add fft helper

Commit:
Untested rough code

Added functions to add optics and calculate output grid.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

MaxThevenet committed:
closer to our use cases

MaxThevenet committed:
Merge branch 'props' into props

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

MaxThevenet committed:
closer to our use cases (#412)

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

MaxThevenet committed:
Merge branch 'props' of https://github.com/LASY-org/lasy into props

Commit:
em-archer

Commit:
em-archer

Commit:
em-archer

Commit:
em-archer

Commit:
em-archer

MaxThevenet committed:
some fixes

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

MaxThevenet committed:
fix axes when len 0

MaxThevenet committed:
fix merge conflict

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

MaxThevenet committed:
well

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

MaxThevenet committed:
typo

MaxThevenet committed:
Merge branch 'props' of https://github.com/LASY-org/lasy into props

MaxThevenet committed:
fix axes

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Commit:
Working Collins

Assumes 'xyt' grids, x and y must be same length and no padding for zoomed-in focus

MaxThevenet committed:
typo

MaxThevenet committed:
fix merge conflicts

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

MaxThevenet committed:
return single element

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

MaxThevenet committed:
return arrays

MaxThevenet committed:
typos

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

MaxThevenet committed:
foo

MaxThevenet committed:
Merge branch 'props' of https://github.com/LASY-org/lasy into props

MaxThevenet committed:
1

MaxThevenet committed:
d

MaxThevenet committed:
test

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

MaxThevenet committed:
fix

MaxThevenet committed:
Merge branch 'props' of https://github.com/LASY-org/lasy into props

MaxThevenet committed:
fix

MaxThevenet committed:
sg

MaxThevenet committed:
fix

MaxThevenet committed:
g

MaxThevenet committed:
rewrite and assert

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

MaxThevenet committed:
fix this issue that was not even in fft

MaxThevenet committed:
Merge branch 'props' of https://github.com/LASY-org/lasy into props

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

MaxThevenet committed:
cleaning

MaxThevenet committed:
fix merge conflicts

MaxThevenet committed:
bit of cleaning in fft

MaxThevenet committed:
renaming

Commit:
Working Collins propagator

There is no padding, currently input grid size and resolution must be tuned manually to match both planes

Commit:
Added chromatic dependencies

Commit:
Cleaning

Commit:
Bug fixing

Commit:
Tidying

Commit:
Cleaning

Commit:
Cleaning

MaxThevenet committed:
Merge branch 'development' into props

MaxThevenet committed:
update only in propagate

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

MaxThevenet committed:
consistency

MaxThevenet committed:
new propagate function

MaxThevenet committed:
cleanup

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

MaxThevenet committed:
copy grid

hightower8083 committed:
[WIP] Props axiprop (#416)

Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>
Co-authored-by: Maxence Thevenet <maxence.thevenet@desy.de>

MaxThevenet committed:
Add docstring to Axiprop propagator, and minor cleaning

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

MaxThevenet committed:
fix docstring formatting

MaxThevenet committed:
abstract

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

MaxThevenet committed:
fix bug in passing arguments to update

MaxThevenet committed:
dummydoc

MaxThevenet committed:
formatting

MaxThevenet committed:
bump up axiprop version dependency

MaxThevenet committed:
cleaning

MaxThevenet committed:
add doc

MaxThevenet committed:
will be added later

MaxThevenet committed:
merge conflicts

MaxThevenet committed:
propagate updates grid.position instead of shifting time axis

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

hightower8083 committed:
a few style fixes in axiprop wrappers doc-strings  (#420)

em-archer committed:
Merge branch 'LASY-org:development' into collins

Commit:
fix merge conflicts

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

MaxThevenet committed:
minor additional comments

MaxThevenet committed:
implement dummy update properly

MaxThevenet committed:
Merge branch 'props' of https://github.com/LASY-org/lasy into props

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

MaxThevenet committed:
missing r

MaxThevenet committed:
Update lasy/propagators/axiprop_propagators.py

Co-authored-by: TimoEichner <55325123+TimoEichner@users.noreply.github.com>

MaxThevenet committed:
fix typos

em-archer committed:
Updated to match #411

MaxThevenet committed:
distance optional so we support abcd

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

MaxThevenet committed:
doc

MaxThevenet committed:
Merge branch 'props' into collins

MaxThevenet committed:
Merge branch 'development' into props

MaxThevenet committed:
Merge branch 'props' into collins

em-archer committed:
Merge branch 'LASY-org:development' into collins

em-archer committed:
Fixing syntax and structure

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

em-archer committed:
Minor fixes

em-archer committed:
Removing units

em-archer committed:
Adding scipy constants

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

em-archer committed:
Major update

- Reformatted structure to mimic existing ASM propagator
- Now utilises the FFT wrapper
- Exception caught if no lens specified

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

em-archer commented:
Have now:
- Restructured to have the same structure as Timo's ASM propagator #421 for readibility.
- Updated to use fft wrapper.
- No longer inherits from single FFT propagator class. A dedicated Collins class for both the single and double methods would be more useful.

Energy conservation is not enforced, because I noticed a strange issue where when calling compute_laser_energy() after initialising a laser pulse, it shows a totally nonsensical number...

<img width="1065" alt="image" src="https://github.com/user-attachments/assets/efd578da-a270-4e79-bb38-1164cec23201" />

MaxThevenet commented on lasy/propagators/collins_sfft_propagator.py:
```diff
@@ -1,9 +1,283 @@
+
+    def add_vacuum(self, distance):
+        vacuum = np.array([[1, distance], [0, 1]])
+        self.abcd = np.matmul(vacuum, self.abcd)
+        return self.abcd
```
Probably no need to return it if the member variable is changed? That applies to several of these functions.

MaxThevenet commented on lasy/propagators/collins_sfft_propagator.py:
```diff
@@ -1,9 +1,283 @@
+        - ``'rt'`` : The laser pulse is represented on a 2D grid:
+                    Cylindrical (r) transversely, and temporal (t) longitudinally.
+
+    abcd : 2d array
+        The 2D ray matrix of the optical system through which the beam propagates.
+        By default, this is initialised to be the unitary matrix:
+
+        .. math::
+
+            O =
+            \begin{pmatrix}
+            A & B \\
+            C & D
+            \end{pmatrix}=
+            \begin{pmatrix}
+            1 & 0 \\
+            0 & 1
+            \end{pmatrix}.
+    """
+
+    def __init__(self, omega0, dim, abcd=np.array([[1, 0], [0, 1]])):
```
The constructor does not need the ABCD, so we can probably do without.

MaxThevenet commented on lasy/propagators/collins_sfft_propagator.py:
```diff
@@ -1,9 +1,283 @@
+
+            r0_step = L0_width / N_points  # Note: D gridpoints means D-1 intervals
+
+            x = fftshift(fftfreq(N_points, r0_step) * lambda0 * f0)
+            y = fftshift(fftfreq(N_points, r0_step) * lambda0 * f0)
```
Do I understand correctly that, here, you need the frequency axis but without doing any FFT? The current fft wrapper implementation does not allow for this, but we could add a function e.g. `fft_axis` or so. Would you then be able to use it?

MaxThevenet commented on lasy/propagators/collins_sfft_propagator.py:
```diff
@@ -1,9 +1,283 @@
+        abcd : array
+            The ray matrix of the optical system
+
+        omega_0 : float (in meter)
+            The wavelength of the electric field
+
+        """
```
Could you add the `Returns` section of this docstring?

MaxThevenet reviewed (commented):
Thanks for updating this PR! See minor comments in the text. Could you also address the automatically generated warnings and suggestions? Also, I would recommend to move the ABCD matrix outside of this propagator, and your implementation is versatile enough already to make this smooth. It would help with clarity, separation of concern and make it visible to other propagators. That would mean creating a new file `lasy/propagators/abcd.py` with a class `ABCD` that has member functions `add_lens` etc., i.e. moving the building of this matrix there. Usage would change as:
### Current use
```py
propagator = CollinsSFFTPropagator(omega0, dim)
propagator.add_lens(f_0) # Add a lens
propagator.add_vacuum(f_0) # Go to focus
propagator.add_vacuum(-3.*150e-6) # A few Rayleigh ranges away
propagator.propagate(laser.grid) # Propagate the pulse upstream of the focal plane
laser.show()
```
### Proposed use (other ways possible)
```py
propagator = CollinsSFFTPropagator(omega0, dim)
abcd = ABCD()
abcd.add_lens(f_0)
abcd.add_vacuum(f_0)
abcd.add_vacuum(-3.*150e-6)
propagator.propagate(laser.grid, abcd)
laser.show()
```
Let me know if you have any questions.

em-archer committed:
Merge branch 'LASY-org:development' into collins

em-archer committed:
Bug fixing for Collins propagator

Moving the ABCD matrix definition to a separate class

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

em-archer committed:
Added tutorial and docstrings

I have now added a tutorial which shows how to do a focal scan for a CW beam. Docstrings have been updated and clarification of returns have been added.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

em-archer committed:
em-archer

em-archer committed:
em-archer

em-archer committed:
em-archer

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

em-archer committed:
em-archer

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

em-archer committed:
em-archer

Better handling of exceptions from incorrect definition of ray matrices

em-archer committed:
em-archer

Changing iPython kernel for tutorial

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

em-archer committed:
em-archer

Updating docstrings

em-archer committed:
em-archer

Fixing docstrings

em-archer committed:
em-archer

em-archer committed:
em-archer

em-archer committed:
Adding tests

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

em-archer committed:
em-archer

Kernel change

em-archer committed:
Fixed tests

Axial ranges needed to be set to be consistent with NF/FF approximations in Collins integral

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

em-archer committed:
Fixing conflicts

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

em-archer committed:
Bug fixing

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

em-archer committed:
Editing docs

em-archer committed:
Fixing docs

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

em-archer committed:
Fixing docs

em-archer committed:
Fixing docs

em-archer committed:
em-archer

em-archer committed:
em-archer

MaxThevenet commented on lasy/propagators/collins_dfft_propagator.py:
```diff
@@ -0,0 +1,302 @@
+                B = self.abcd.abcd[0][1]
+                C = self.abcd.abcd[1][0]
+                D = self.abcd.abcd[1][1]
+                det = A * D - B * C
```
Thanks!! Could you look at these suggestions from our linter?

em-archer replied:
Hi Maxence, would you be able to help me implement this properly? This is what I am trying to do:

<img width="362" alt="image" src="https://github.com/user-attachments/assets/ef218bd3-8147-4052-8943-deae6bf0f9c2" />

em-archer committed:
Tidying

em-archer committed:
Reverting temporarily to SFFT

Reverting to SFFT only until I have more time to debug tests for DFFT

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

em-archer committed:
Changing to basic Python kernel for tutorial

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

em-archer committed:
Fixing energy conservation

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

MaxThevenet commented on docs/source/tutorials/collins_propagator.ipynb:
```diff
@@ -0,0 +1,266 @@
+   "id": "9296ccf6-4e83-467f-afe5-a74e6820e110",
+   "metadata": {},
+   "source": [
+    "# Performing a focal scan using geometric optics and diffraction with the Collins propagator"
```
```suggestion
    "# Collins propagator and focal scan"
```

MaxThevenet commented on docs/source/tutorials/collins_propagator.ipynb:
```diff
@@ -0,0 +1,266 @@
+   "id": "cbd257fa-9f45-4ab1-b338-cf84ccc52119",
+   "metadata": {},
+   "source": [
+    "We will now make a copy of the laser and define both the SFFT and DFFT propagators that we will use in our focal scan. We then initialise the ABCD matrix and add a thin-lens of the correct focal length:"
```
```suggestion
    "We will now make a copy of the laser and define both the SFFT propagator that we will use in our focal scan. We then initialise the ABCD matrix and add a thin-lens of the correct focal length:"
```

MaxThevenet committed:
Merge branch 'development' into collins

MaxThevenet committed:
minor cleaning in Collins tutorial

MaxThevenet committed:
no need of ABCD.update

MaxThevenet committed:
minor style checks

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

MaxThevenet reviewed (approved):
Great, thanks for this PR!!
